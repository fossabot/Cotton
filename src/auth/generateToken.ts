import { readFileSync } from 'fs';
import * as jwt from 'jsonwebtoken';

function jwtDate(date: Date) {
  return Math.floor(date.getTime() / 1000);
}

// Generate JWT for GitHub App.
// Requires gh_priv_key.pem to be in repo root (i.e. ../..).
// Token docs: https://developer.github.com/apps/building-github-apps/authentication-options-for-github-apps/#authenticating-as-a-github-app
export default function generateGitHubToken() {
  const cert = readFileSync('gh_priv_key.pem');

  const payload = {
    iat: jwtDate(new Date()),
    exp: jwtDate(new Date()) + 100,
    iss: 10823,
  };

  const options = {
    algorithm: 'RS256',
  };

  return jwt.sign(payload, cert, options);
}
