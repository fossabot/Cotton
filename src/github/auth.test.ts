import { generateGitHubToken, signRequestBody, verifyWebhookEvent } from './auth';
import { APIGatewayEvent } from 'aws-lambda';
import * as jwt from 'jsonwebtoken';
import * as _ from 'lodash';

jest.mock('fs');
const { readFileSync } = require('fs');

const mockPublicKey = `-----BEGIN PUBLIC KEY-----
MFwwDQYJKoZIhvcNAQEBBQADSwAwSAJBAIR7xbGOHtkjr1B0hc9ZxglnHKbMfySC
ccOjOOxiF7xznmmbYY1N5SKnSkwX7zKv+JR3Gk9wbFB7xujvpP3NSq0CAwEAAQ==
-----END PUBLIC KEY-----
`;

const mockPrivateKey = `
-----BEGIN RSA PRIVATE KEY-----
MIIBOgIBAAJBAIR7xbGOHtkjr1B0hc9ZxglnHKbMfySCccOjOOxiF7xznmmbYY1N
5SKnSkwX7zKv+JR3Gk9wbFB7xujvpP3NSq0CAwEAAQJAXtIfipdHqO4LejAL3X5J
aU1tikxu63iZaAaYpUcH6g5l3zBW6GwBNGrtmv/8XrCOmDoNOBpcjpklNUE0wTZW
mQIhAMflk8vTwAoSKpvaVVfPZM5FdMjKCfiLOpJL5Cu3wRDbAiEAqaqaifyPH7su
nS/8YZT5xc5KYp/wHT/b0B+oJKIJhRcCIQCrBKHmnznMI/CqOCgNoRSoWMnqQtg2
+L7ajO0f7ezSQwIgb5Uoz4kPI8I51NzDMfYFEocqOpSPGN1vlf/L2FPMyP8CIGiE
yNvg2ztwi3BKf6ufRqi0Ma8bL3Sy5jls/f472mzI
-----END RSA PRIVATE KEY-----
`;

describe(generateGitHubToken, () => {
  test('should meet GitHub specs', () => {
    readFileSync.mockReturnValue(mockPrivateKey);
    const token = generateGitHubToken();
    readFileSync.mockReset();

    // Expect token to be signed correctly
    expect(() => jwt.verify(token, mockPublicKey)).not.toThrow();

    // Expect token to be signed with the RS256 algorithm.
    // GitHub requirement.
    const decoded = jwt.decode(token, { complete: true });
    expect(decoded).toMatchObject({ header: { alg: 'RS256' } });
  });
});

describe(signRequestBody, () => {
  test('should return null if body does not exist', () => {
    expect(signRequestBody('key', null)).toBe(null);
  });

  test('should return signed body', () => {
    expect(signRequestBody('key', 'body')).toMatchSnapshot();
  });
});

describe(verifyWebhookEvent, () => {
  beforeEach(() => {
    process.env.GITHUB_WEBHOOK_SECRET = '';
  });

  afterEach(() => {
    delete process.env.GITHUB_WEBHOOK_SECRET;
  });

  const validEvent: APIGatewayEvent = {
    body: '',
    headers: {
      'X-Hub-Signature': 'sig',
      'X-GitHub-Event': 'event',
      'X-GitHub-Delivery': 'delivery',
    },
  } as any;

  function expectValidError(event: APIGatewayEvent) {
    const error = verifyWebhookEvent(event);
    expect(error).toBeDefined();
    expect(Object.keys(error || {})).toEqual(['statusCode', 'body']);
  }

  test('should return error when webhook secret is not set', () => {
    delete process.env.GITHUB_WEBHOOK_SECRET;
    expectValidError(validEvent);
  });

  function testMissingHeader(header: string) {
    test(`should return error when ${header} header is not set`, () => {
      const event = _.cloneDeep(validEvent);
      delete event.headers[header];
      expectValidError(event);
    });
  }

  testMissingHeader('X-Hub-Signature');
  testMissingHeader('X-GitHub-Event');
  testMissingHeader('X-GitHub-Delivery');

  test('should return error when signature is invalid', () => {
    // Signature set doesn't match signature calculated by signRequestBody
    expectValidError(validEvent);
  });

  test.skip('should return null when webhook event is valid', () => {
    // TODO: Mock signRequestBody somehow to get past that last signature check
    expect(verifyWebhookEvent(validEvent)).toBe(null);
  });
});
