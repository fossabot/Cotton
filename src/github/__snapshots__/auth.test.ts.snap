// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signRequestBody should return signed body 1`] = `"sha1=70bbf6819d1037aa94ca7e7f537cbea25fe49283"`;
